Firestore/core/src/firebase/firestore/model/transform_operations.h

```diff
@@ -233,6 +245,116 @@ class ArrayTransform : public TransformOperation {
   }
 };
 
+/**
+ * Implements the backend semantics for locally computed NUMERIC_ADD (increment)
+ * transforms. Converts all field values to longs or doubles and resolves
+ * overflows to LONG_MAX/LONG_MIN.
+ */
+class NumericIncrementTransform : public TransformOperation {
+ public:
+  NumericIncrementTransform(FSTNumberValue* operand)
+      : operand_(std::move(operand)) {
+  }
+
+  ~NumericIncrementTransform() override {
+  }
+
+  Type type() const override {
+    return Type::Increment;
+  }
+
+  FSTFieldValue* ApplyToLocalView(
+      FSTFieldValue* previousValue,
+      FIRTimestamp* /* localWriteTime */) const override {
+    // Return an integer value only if the previous value and the operand is an
+    // integer.
+    if ([previousValue isKindOfClass:[FSTIntegerValue class]] &&
+        [operand_ isKindOfClass:[FSTIntegerValue class]]) {
+      int64_t sum = SafeIncrement(
+          (reinterpret_cast<FSTIntegerValue*>(previousValue)).internalValue,
```

Comment:
Under C++ rules, `static_cast` would be sufficient throughout the function (it is valid to `static_cast` a base class to a derived class). Unless it doesn't play well with Objective-C, `static_cast` is preferable -- in general, [it is best to use the least powerful cast applicable](https://stackoverflow.com/a/332086).